File: src/main/java/HGTM/Groupform.java
package HGTM;

import input.Group;
import input.Agent;
import main.Function;
import org.jgrapht.alg.interfaces.ShortestPathAlgorithm;
import org.jgrapht.graph.DefaultUndirectedWeightedGraph;
import org.jgrapht.graph.DefaultWeightedEdge;

import java.util.*;

/**
 * @author Liu tianyi
 */
public class Groupform {
    private DefaultUndirectedWeightedGraph<Integer, DefaultWeightedEdge> arcGraph;
    private HashMap<Integer, Group> idToGroups;
    private HashMap<Integer, Agent> idToAgents;
    private ShortestPathAlgorithm<Integer, DefaultWeightedEdge> shortestPath ;
    Double a;
    Double b;
    List<List<Agent>> bags = new ArrayList<>();
    Map<List<Agent>,Agent> bagsToAgent = new HashMap<>();

    public Groupform(DefaultUndirectedWeightedGraph<Integer, DefaultWeightedEdge> arcGraph, HashMap<Integer, Group> idToGroups,
                     HashMap<Integer, Agent> idToAgents, ShortestPathAlgorithm<Integer, DefaultWeightedEdge> shortestPath, Double a, Double b) {
        this.arcGraph = arcGraph;
        this.idToGroups = idToGroups;
        this.idToAgents = idToAgents;
        this.shortestPath = shortestPath;
        this.a = a;
        this.b = b;
        for (Integer id : idToAgents.keySet()) {
            List<Agent> bag =new ArrayList<>();
            Agent e = idToAgents.get(id);
            if (e.getFaultA()==1) {
                bag.add(e);
                bags.add(bag);
            }
        }
    }
    public Map<List<Agent>,Agent> run() {
        IntraBagform();
        //InterBagform();
        Map<List<Agent>,Agent> bagsToAgentReturn = new HashMap<>();
        for (List<Agent> bag : bags) {
            bagsToAgentReturn.put(bag,bagsToAgent.get(bag));
        }
        return bagsToAgentReturn;
    }

    Comparator<List<Agent>> CompareBag= new Comparator<List<Agent>>() {
        @Override
        public int compare(List<Agent> o1, List<Agent> o2) {
            int size1 = 0;
            int size2 = 0;
            for (Agent agent : o1) {
                size1+=agent.getTasksList().size();
            }
            for (Agent agent : o2) {
                size2+=agent.getTasksList().size();
            }

            return size2-size1;
        }
    };


    private void IntraBagform(){
        //进行网络层内的组件打包
        PriorityQueue<List<Agent>> pq = new PriorityQueue<>(CompareBag);
        pq.addAll(bags);
        List<List<Agent>> temp = new ArrayList<>();
        //用来容纳 无法再无法再通过合并的方式扩大的任务组
        while (!pq.isEmpty()) {
            List<Agent> bagM = pq.poll();
            int flag = 0;
            PriorityQueue<List<Agent>> pqTemp = new PriorityQueue<>(pq);
            int size = pqTemp.size();
            //flag 和size用于判断 是否该包已经无法再通过合并的方式
            while (!pqTemp.isEmpty()) {
                List<Agent> bagN = pqTemp.poll();
                if (bagN==null) {
                    continue;
                }
                List<Agent> bagTemp = new ArrayList<>(bagN);
                bagTemp.addAll(bagM);
                if (BenIntra(bagTemp)>(BenIntra(bagN)+ BenIntra(bagM))) {
                    pq.offer(bagTemp);
                    bagsToAgent.remove(bagM);
                    bagsToAgent.remove(bagN);
                } else {
                    flag++;
                }
            }
            if (flag==size) {
                temp.add(bagM);
            }
        }
        bags = new ArrayList<>(temp);
    }

    private void InterBagform() {
        //进行网络层间的组件打包
        PriorityQueue<List<Agent>> pq = new PriorityQueue<>(CompareBag);
        pq.addAll(bags);
        List<List<Agent>> temp = new ArrayList<>();
        //用来容纳 无法再无法再通过合并的方式扩大的任务组
        while (!pq.isEmpty()) {
            List<Agent> bagM = pq.poll();
            int flag = 0;
            PriorityQueue<List<Agent>> pqTemp = new PriorityQueue<>(pq);
            int size = pqTemp.size();
            //flag 和size用于判断 是否该包已经无法再通过合并的方式
            while (!pqTemp.isEmpty()) {
                List<Agent> bagN = pqTemp.poll();
                if (bagN==null) {
                    continue;
                }
                List<Agent> bagTemp = new ArrayList<>(bagN);
                bagTemp.addAll(bagM);
                if (BenInter(bagTemp)>(BenInter(bagN)+ BenInter(bagM))) {
                    pq.offer(bagTemp);
                    bagsToAgent.remove(bagM);
                    bagsToAgent.remove(bagN);
                }else {
                    flag++;
                }
                if (flag==size) {
                    temp.add(bagM);
                }
            }
        }
        bags = new ArrayList<>(temp);
    }

    private Double BenIntra(List<Agent> bagTemp) {
        Double benIntraValue = Double.MIN_VALUE;
        List<Agent> neighbors = new ArrayList<>();
        for (Agent agent : bagTemp) {
            Set<DefaultWeightedEdge> defaultWeightedEdges = arcGraph.edgesOf(agent.getRobotId());
            for (DefaultWeightedEdge defaultWeightedEdge : defaultWeightedEdges) {
                Agent e = idToAgents.get(arcGraph.getEdgeTarget(defaultWeightedEdge));
                if (e.getGroupId()!=agent.getGroupId()||e.getFaultA()==1) {
                    continue;
                }
                neighbors.add(e);
            }
        }
        Agent target = null;
        for (Agent neighbor : neighbors) {
            double benIntraTemp = benIntra(bagTemp, neighbor);
            if (benIntraTemp >benIntraValue) {
                benIntraValue  = benIntraTemp;
                target = neighbor;
            }
        }
        bagsToAgent.put(bagTemp,target);
        return benIntraValue;
    }
    /*
       计算将任务包bagTemp同层迁移给临近节点neighbor的收益函数
     */
    private Double benIntra(List<Agent> bagTemp, Agent neighbor) {
        Group group = idToGroups.get(neighbor.getGroupId());
        double interactionLevel = group.getInteractionLevel();
        double meanC = 0.0;
        //任务完成成本增加比例的分子部分
        Double CostDenominator = 0.0;
        //任务完成成本增加比例的分母部分
        int count = 0;
        Set<DefaultWeightedEdge> defaultWeightedEdges = arcGraph.edgesOf(neighbor.getRobotId());
        for (DefaultWeightedEdge defaultWeightedEdge : defaultWeightedEdges) {
            Agent e = idToAgents.get(arcGraph.getEdgeTarget(defaultWeightedEdge));
            if (e.getGroupId()!=neighbor.getGroupId()) {
                continue;
            }
            CostDenominator+=arcGraph.getEdgeWeight(defaultWeightedEdge)*e.getTasksList().size();
            count++;
            //count用来表示目标agent neighbor有多少个临近节点
            meanC+=e.getLoad()/e.getCapacity();
        }
        CostDenominator/=count;
        Double loadInBag = 0.0;
        for (Agent agent : bagTemp) {
            CostDenominator+=agent.getLoad();
            loadInBag+=agent.getLoad();
        }

        for (Agent agent : bagTemp) {
            DefaultWeightedEdge weightedEdge = arcGraph.getEdge(agent.getRobotId(), neighbor.getRobotId());
            if (weightedEdge==null) {
                continue;
            }
            CostDenominator+=arcGraph.getEdgeWeight(weightedEdge);
        }
        Double CostIncreaseP = CostDenominator/meanC;
        Function function = new Function(idToAgents,idToGroups);


        Double CompleteP = 1-Math.max(function.sig(loadInBag) * interactionLevel, 0.5);
        return b*CompleteP-a*CostIncreaseP;

    }



    private Double BenInter(List<Agent> bagTemp) {
        Double benInterValue = Double.MIN_VALUE;
        List<Agent> neighbors = new ArrayList<>();
        for (Agent agent : bagTemp) {
            Set<DefaultWeightedEdge> defaultWeightedEdges = arcGraph.edgesOf(agent.getRobotId());
            for (DefaultWeightedEdge defaultWeightedEdge : defaultWeightedEdges) {
                Agent e = idToAgents.get(arcGraph.getEdgeTarget(defaultWeightedEdge));
                if (e.getGroupId()!=agent.getGroupId()) {
                    continue;
                }
                neighbors.add(e);
            }
        }
        Agent target = new Agent();
        for (Agent neighbor : neighbors) {
            double benIntraTemp = benInter(bagTemp, neighbor);
            if (benIntraTemp >benInterValue) {
                benInterValue  = benIntraTemp;
            }
        }
        bagsToAgent.put(bagTemp,target);
        return benInterValue;
    }
/*
   计算将任务包bagTemp跨层迁移给临近节点neighbor的收益函数
 */
    private Double benInter(List<Agent> bagTemp, Agent neighbor) {
        Double meanC = 0.0;
        //任务完成成本增加比例的分子部分
        Double CostDenominator = 0.0;
        //任务完成成本增加比例的分母部分
        int count = 0;
        Set<DefaultWeightedEdge> defaultWeightedEdges = arcGraph.edgesOf(neighbor.getRobotId());
        for (DefaultWeightedEdge defaultWeightedEdge : defaultWeightedEdges) {
            Agent e = idToAgents.get(arcGraph.getEdgeTarget(defaultWeightedEdge));
            CostDenominator+=arcGraph.getEdgeWeight(defaultWeightedEdge)*e.getTasksList().size();
            count++;
            //count用来表示目标agent neighbor有多少个临近节点
            meanC+=e.getLoad()/e.getCapacity();
        }
        CostDenominator/=count;
        Double loadInBag = 0.0;
        for (Agent agent : bagTemp) {
            CostDenominator+=agent.getLoad();
            loadInBag+=agent.getLoad();
        }

        for (Agent agent : bagTemp) {
            DefaultWeightedEdge weightedEdge = arcGraph.getEdge(agent.getRobotId(), neighbor.getRobotId());
            if (weightedEdge==null) {
                continue;
            }
            CostDenominator+=arcGraph.getEdgeWeight(weightedEdge);
        }
        Double CostIncreaseP = CostDenominator/meanC;
        Function function = new Function(idToAgents,idToGroups);


        Double CompleteP = 1-Math.max(function.sig(loadInBag), 0.5);
        return b*CompleteP-a*CostIncreaseP;
    }
}

File: python_src/hgtm/groupform.py
"""Groupform class for grouping agents into bags"""
import sys
sys.path.append('..')
from python_src.main.function import Function


class Groupform:
    def __init__(self, arc_graph, id_to_groups, id_to_agents, a, b):
        self.arc_graph = arc_graph
        self.id_to_groups = id_to_groups
        self.id_to_agents = id_to_agents
        self.a = a
        self.b = b
        self.bags = []
        self.bags_to_agent = {}

        # Initialize bags with faulty agents
        for agent_id in id_to_agents.keys():
            e = id_to_agents[agent_id]
            if e.get_fault_a() == 1:
                bag = [e]
                self.bags.append(bag)

    def run(self):
        """Run groupform algorithm"""
        self.intra_bagform()
        bags_to_agent_return = {}

        for bag in self.bags:
            bag_tuple = tuple(bag)
            if bag_tuple in self.bags_to_agent:
                bags_to_agent_return[bag_tuple] = self.bags_to_agent[bag_tuple]

        return bags_to_agent_return

    def compare_bag(self, bag):
        """Compare bags by total task count"""
        size = sum(len(agent.get_tasks_list()) for agent in bag)
        return -size  # Negative for descending order

    def intra_bagform(self):
        """Perform intra-layer bag formation"""
        import heapq

        pq = [(self.compare_bag(bag), id(bag), bag) for bag in self.bags]
        heapq.heapify(pq)

        temp = []

        while pq:
            _, _, bag_m = heapq.heappop(pq)
            merged = False

            # Extract all remaining bags from pq
            remaining_bags = []
            while pq:
                remaining_bags.append(heapq.heappop(pq))

            # Try to merge bag_m with each remaining bag
            for i, (priority, bag_id, bag_n) in enumerate(remaining_bags):
                if bag_n is None:
                    continue

                bag_temp = list(bag_n) + list(bag_m)

                ben_temp = self.ben_intra(bag_temp)
                ben_n = self.ben_intra(bag_n)
                ben_m = self.ben_intra(bag_m)

                if ben_temp > (ben_n + ben_m):
                    # Merge is beneficial
                    # Remove old entries from bags_to_agent
                    bag_m_tuple = tuple(bag_m)
                    bag_n_tuple = tuple(bag_n)
                    if bag_m_tuple in self.bags_to_agent:
                        del self.bags_to_agent[bag_m_tuple]
                    if bag_n_tuple in self.bags_to_agent:
                        del self.bags_to_agent[bag_n_tuple]

                    # Put merged bag back in pq
                    heapq.heappush(pq, (self.compare_bag(bag_temp), id(bag_temp), bag_temp))

                    # Put back all other bags except bag_n (which was merged)
                    for j, bag_item in enumerate(remaining_bags):
                        if j != i:
                            heapq.heappush(pq, bag_item)

                    merged = True
                    break

            if not merged:
                # No beneficial merge found, bag_m is final
                temp.append(bag_m)
                # Put back all remaining bags
                for bag_item in remaining_bags:
                    heapq.heappush(pq, bag_item)

        self.bags = temp

    def ben_intra(self, bag_temp):
        """Calculate benefit of intra-layer migration"""
        ben_intra_value = -float('inf')
        neighbors = []

        for agent in bag_temp:
            edges = self.arc_graph.edges(agent.get_robot_id())

            for edge in edges:
                if edge[0] == agent.get_robot_id():
                    target_id = edge[1]
                else:
                    target_id = edge[0]

                e = self.id_to_agents[target_id]

                if e.get_group_id() != agent.get_group_id() or e.get_fault_a() == 1:
                    continue

                neighbors.append(e)

        target = None

        for neighbor in neighbors:
            ben_intra_temp = self.ben_intra_calc(bag_temp, neighbor)

            if ben_intra_temp > ben_intra_value:
                ben_intra_value = ben_intra_temp
                target = neighbor

        if target is not None:
            self.bags_to_agent[tuple(bag_temp)] = target

        return ben_intra_value

    def ben_intra_calc(self, bag_temp, neighbor):
        """Calculate benefit for specific neighbor"""
        group = self.id_to_groups[neighbor.get_group_id()]
        interaction_level = group.get_interaction_level()
        mean_c = 0.0
        cost_denominator = 0.0
        count = 0

        edges = self.arc_graph.edges(neighbor.get_robot_id())

        for edge in edges:
            if edge[0] == neighbor.get_robot_id():
                target_id = edge[1]
            else:
                target_id = edge[0]

            e = self.id_to_agents[target_id]

            if e.get_group_id() != neighbor.get_group_id():
                continue

            weight = self.arc_graph[edge[0]][edge[1]]['weight']
            cost_denominator += weight * len(e.get_tasks_list())
            count += 1
            mean_c += e.get_load() / e.get_capacity()

        if count > 0:
            cost_denominator /= count

        load_in_bag = sum(agent.get_load() for agent in bag_temp)
        cost_denominator += load_in_bag

        for agent in bag_temp:
            edge_data = self.arc_graph.get_edge_data(agent.get_robot_id(),
                                                     neighbor.get_robot_id())
            if edge_data is not None:
                cost_denominator += edge_data['weight']

        cost_increase_p = cost_denominator / mean_c if mean_c > 0 else 0

        function = Function(self.id_to_agents, self.id_to_groups)
        complete_p = 1 - max(function.sig(load_in_bag) * interaction_level, 0.5)

        return self.b * complete_p - self.a * cost_increase_p
